### PR TITLE
multitenant: Extend retry for upgrade rpc

### LINKED ISSUE #114004: roachtest: multitenant-upgrade failed
roachtest.multitenant-upgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12550589?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/12550589?buildTab=artifacts#/multitenant-upgrade) on release-23.2 @ [56fd045302ed2edc14831fb001ef77486ae8f62a](https://github.com/cockroachdb/cockroach/commits/56fd045302ed2edc14831fb001ef77486ae8f62a):

```
(sql_runner.go:98).ExecWithMessage: error executing query="SET CLUSTER SETTING version = crdb_internal.node_executable_version()" args=[]: pq: upgrade failed due to transient SQL servers
(cluster.go:2220).Run: context canceled
(cluster.go:2220).Run: context canceled
(cluster.go:2220).Run: context canceled
(cluster.go:2220).Run: context canceled
test artifacts and logs in: /artifacts/multitenant-upgrade/run_1
```
<p>Parameters: <code>ROACHTEST_arch=amd64</code>
, <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_fs=ext4</code>
, <code>ROACHTEST_localSSD=true</code>
, <code>ROACHTEST_metamorphicBuild=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-12550589/multitenant-upgrade/1699423627127/1699423834766)

</p>
</details>
/cc @cockroachdb/multi-tenant
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*multitenant-upgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-33291

### PR DESCRIPTION
**Pull request by ajstorm**
The tenant upgrade interlock requires contacting each active SQL server during a tenant upgrade. In cases where a remote SQL server stops and starts rapidly during (or before) and upgrade, the RPC connection will get severed and reconnected. Currently we don't retry long enough to catch this reconnection in all cases. This commit extends the retry to a full second to ensure that we don't fail the upgrade prematurely.

Fixes: #114004
Release note: None
Release justification: Low risk change to address possible customer issue

### CODE REVIEW

File: pkg/upgrade/upgradecluster/tenant_cluster.go [hunk 1]
```diff
@@ -232,7 +233,7 @@ func (t *TenantCluster) ForEveryNodeOrServer(
 			var conn *grpc.ClientConn
 			retryOpts := retry.Options{
 				InitialBackoff: 0,
-				MaxRetries:     2,
+				MaxRetries:     100,
 				MaxBackoff:     10 * time.Millisecond,
```
Comment (stevendanna) [inline pkg/upgrade/upgradecluster/tenant_cluster.go:237]: What do you think about maybe 20 as the max retries but then push the MaxBackoff to 100ms?

Comment (ajstorm) [reply pkg/upgrade/upgradecluster/tenant_cluster.go]: Yeah, that makes more sense.  I've also increased the initial backoff to 1 millisecond, to match other uses of retry.Options.

**Review by yuzefovich — APPROVED** [review]
:lgtm:

The issue seems to be test-only (based on the comment from Ahmad added in 1f6e7341c9ad2833ff08533c46b3106af4c84ccf), so should this not be a beta blocker?

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114397#-:-NjDdpvcDPIqYsKSErlu:bzasfj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm and @stevendanna)

___
*[`pkg/upgrade/upgradecluster/tenant_cluster.go` line 255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114397#-NjDdRruCkv0zlecIJcD:-NjDdRruCkv0zlecIJcE:bcy8isq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8807ee3b759c9951e884adfd2049fb49d3c2c71/pkg/upgrade/upgradecluster/tenant_cluster.go#L255)):*
> ```Go
> }
> 
> func annotateDialError(err error, id base.SQLInstanceID) error {
> ```

nit: did you intend to use `id` in an error message somewhere?

**Review by ajstorm — COMMENTED** [review]
Thanks for the review.  Yes, this should not be a release blocker.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114397#-:-NjE6UXv26-AHKoFNMfo:b-rhoqr0)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna and @yuzefovich)

___
*[`pkg/upgrade/upgradecluster/tenant_cluster.go` line 255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114397#-NjDdRruCkv0zlecIJcD:-NjE6MgTA1bu7jxZrBFd:bqdot35) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8807ee3b759c9951e884adfd2049fb49d3c2c71/pkg/upgrade/upgradecluster/tenant_cluster.go#L255)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: did you intend to use `id` in an error message somewhere?
</blockquote></details>

Good catch.  I've removed it.

**Review by yuzefovich — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114397)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna)

**Comment by ajstorm** [comment]
bors r=yuzefovich

**Comment by yuzefovich** [comment]
blathers backport 23.2

**Pull request merged**